### PR TITLE
feat: Append job creation timestamp to cronjob annotations

### LINKED ISSUE #108609: Inject schedule name into newly created job
### What happened?

In cronjob_controllerv2.go, jobs name pass from epoch in seconds to epoch in minutes

`name := fmt.Sprintf("%s-%d", cj.Name, getTimeHash(scheduledTime))`

to 

`
func getJobName(cj *batchv1.CronJob, scheduledTime time.Time) string {
	return fmt.Sprintf("%s-%d", cj.Name, getTimeHashInMinutes(scheduledTime))
}
`

### What did you expect to happen?

Expected this to be documented somewhere.

This is a breaking change, affecting a known strategy to obtain the job expected runtime. Eg. https://stackoverflow.com/questions/55694707/access-scheduled-run-time-inside-kubertes-cronjob#comment120366156_55694854

### How can we reproduce it (as minimally and precisely as possible)?

-

### Anything else we need to know?

-

### Kubernetes version

<details>

Release 1.23 removed old controller.

</details>

### Cloud provider

<details>

</details>

### OS version

_No response_

### Install tools

_No response_

### Container runtime (CRI) and version (if applicable)

_No response_

### Related plugins (CNI, CSI, ...) and versions (if applicable)

_No response_

### PR DESCRIPTION
**Pull request by helayoty**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
- Add new annotation `CronJobScheduledTimestampAnnotation` to the batch API.
- Update features with `CronJobsScheduledAnnotation`.
- Pass context to the `getJobFromTemplate2` func to be able to log the errors.
- Update Unit test.

#### Which issue(s) this PR fixes:
Fixes #108609

#### Does this PR introduce a user-facing change? 
Yes.

```release-note
Add new annotation `batch.kubernetes.io/cronjob-scheduled-timestamp` to Job objects scheduled from CronJobs.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
- https://github.com/kubernetes/enhancements/issues/4026
- https://github.com/kubernetes/enhancements/pull/4027 

```docs
```

### CODE REVIEW

File: pkg/controller/cronjob/utils_test.go
Comment (troy0820) [inline pkg/controller/cronjob/utils_test.go:94]: Unit test seems to be failing as the scheduledTime isn't what is set on the above appened func but the creationtimestamp of the batch job.  

The scheduledTime in seconds gets appended to the name of the job but the creationTimestamp seem to be different than what is passed in.

File: pkg/controller/cronjob/utils.go
Comment (troy0820) [inline pkg/controller/cronjob/utils.go:194]: Do we want to use the scheduledTime as the String to the appended annotation?  This is also being set (appended) as the name from the getJobname which will ensure consistency.

File: pkg/controller/cronjob/utils.go
Comment (troy0820) [inline pkg/controller/cronjob/utils.go:193]: nit:

Comment suggests that we are adding the name to the job annotation where  we are adding the timestamp to the annotation

File: pkg/controller/cronjob/utils_test.go
Comment (troy0820) [reply pkg/controller/cronjob/utils_test.go:94]: So the Annotation isn't going to equal the string of the scheduledTime being that this will be different than what is set when creating the job.

**Review by troy0820 — COMMENTED** [review]
Some nits and questions around if we want the timestamp from creation or the scheduledtime that gets appended to the job name.  

If we want the former, then the test associated with this change is going to fail as the creationTimeStamp of the job and not the scheduledTime passed into the `getJobFromTemplate2` function will be different.

File: pkg/controller/cronjob/utils_test.go
Comment (troy0820) [reply pkg/controller/cronjob/utils_test.go:94]: This change should work.

File: pkg/controller/cronjob/utils.go
Comment (helayoty) [reply pkg/controller/cronjob/utils.go]: Thanks for the catch. Fixed.

File: pkg/controller/cronjob/utils_test.go
Comment (helayoty) [reply pkg/controller/cronjob/utils_test.go:94]: The annotation will have the original scheduled time, the `scheduledTime` will be set as the new value for the`CreationTimestamp` meta and the user doesn't have a way to get the original/expected scheduled time.

File: pkg/controller/cronjob/utils.go
Comment (helayoty) [reply pkg/controller/cronjob/utils.go:194]: It might confuse others to use the scheduledTime instead, as we will append the expected/initial CreationTime.

**Comment by helayoty** [comment]
> Some nits and questions around if we want the timestamp from creation or the scheduledtime that gets appended to the job name.
> 
> If we want the former, then the test associated with this change is going to fail as the creationTimeStamp of the job and not the scheduledTime passed into the `getJobFromTemplate2` function will be different.

@troy0820  Thank you for your review. The annotation will have the original scheduled time, the scheduledTime will be set as the new value for theCreationTimestamp meta, and the user doesn't have a way to get the original/expected scheduled time.

File: pkg/controller/cronjob/utils.go
Comment (soltysh) [inline pkg/controller/cronjob/utils.go:212]: A few comments here:
1. I don't think we need entire new function just for adding one more annotation. I'd leave this line
```golang
annotations := copyAnnotations(&cj.Spec.JobTemplate)
```
which ensures the set exists, and then just:
```golang
annotations[batchv1.JobCreationTimestampAnnotation] = <time>
```
2. `creationTimestamp.String()` isn't safe since it may change, and since we expose this information as API we need to ensure the stability of the data. I'd suggest having an explicit date-time format: `YYYY-MM-DD HH:MM`, for example. You also need to be explicit about what timezone this value is in, since we also support that, so it's reasonable to think through which one we want to present here.

File: staging/src/k8s.io/api/batch/v1/types.go
Comment (soltysh) [inline staging/src/k8s.io/api/batch/v1/types.go:30]: Make sure to explain in details the purpose of this annotation, its value and expectations user should have wrt to its format and its contents.

**Review by soltysh — CHANGES_REQUESTED** [review]
Also based on the discussion in https://github.com/kubernetes/kubernetes/issues/117986 I'm inclined to say that this particular change might also require a KEP, a small one, similar to https://github.com/kubernetes/enhancements/pull/4019 but still would be good to describe the format of the date and ensure this is working as expected. 
/hold

File: staging/src/k8s.io/api/batch/v1/types.go
Comment (helayoty) [reply staging/src/k8s.io/api/batch/v1/types.go:32]: Added.

File: pkg/controller/cronjob/utils.go
Comment (helayoty) [reply pkg/controller/cronjob/utils.go]: Both comments have been addressed.

File: staging/src/k8s.io/api/batch/v1/types.go
Comment (sftim) [inline staging/src/k8s.io/api/batch/v1/types.go:28]: BTW, this prefix is (already) used for both labels and annotations.

**Review by sftim — COMMENTED** [review]
BTW, if this merges we should also send in a PR for k/website on the _dev-1.28_ branch, to ensure we update https://kubernetes.io/docs/reference/labels-annotations-taints/

File: staging/src/k8s.io/api/batch/v1/types.go
Comment (helayoty) [reply staging/src/k8s.io/api/batch/v1/types.go:28]: Do we need to create a separate prefix for annotation?

Comment (sftim) [reply staging/src/k8s.io/api/batch/v1/types.go:28]: We can use the same prefix. It's the name of the const that we might want to tweak.

File: pkg/controller/cronjob/utils_test.go
Comment (helayoty) [reply pkg/controller/cronjob/utils_test.go:41]: Addressed.

**Comment by helayoty** [comment]
> BTW, if this merges we should also send in a PR for k/website on the _dev-1.28_ branch, to ensure we update https://kubernetes.io/docs/reference/labels-annotations-taints/

@sftim, thanks for the review and for the hint. I've added it as part of the doc deliverable in the KEP [issue](https://github.com/kubernetes/enhancements/issues/4026).

File: pkg/controller/cronjob/utils_test.go
Comment (sftim) [inline pkg/controller/cronjob/utils_test.go:92]: I think these thee strings are RFC3339 strings that represent the same instant:
- `2023-05-34T23:39:57-00:00`
- `2023-05-34T23:39:57+00:00`
- `2023-05-34T23:39:57Z`

However, we'd like to be consistent between our releases. Using a fixed string here would make it obvious to any reviewer when we're changing the outcome. That's why I recommend it.

Testing that `Format` returns the same result as another `Format` doesn't feel as helpful.

**Comment by furkatgofurov7** [comment]
@helayoty hey 👋🏼 !

Just a reminder that the code freeze is starting [01:00 UTC Wednesday 19th July 2023 / 18:00 PDT Tuesday 18th July 2023](https://everytimezone.com/s/72ee8496) (about 3 weeks from now) and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.

File: pkg/controller/cronjob/utils.go
Comment (soltysh) [inline pkg/controller/cronjob/utils.go:220]: nit: default will match the kube-controller-manager time zone.

File: staging/src/k8s.io/api/batch/v1/types.go
Comment (soltysh) [inline staging/src/k8s.io/api/batch/v1/types.go:30]:
```suggestion
	// CronJobScheduledTimestampAnnotation is a creation timestamp annotation for the Job.
```

**Review by soltysh — APPROVED** [review]
/approve
the controller changes, module minor nits

**Comment by soltysh** [comment]
/label api-review 
for the new annotation, only

**Review by soltysh — APPROVED** [review]
/lgtm
/approve
from sig-apps pov

File: pkg/controller/cronjob/utils.go
Comment (soltysh) [inline pkg/controller/cronjob/utils.go:228]: While reviewing https://github.com/kubernetes/website/pull/41857 I've noticed this bit, since in the controller code we will [not start a job with invalid time zone](https://github.com/kubernetes/kubernetes/blob/2f563464bc879cc7d31642eb77e2b0d38b498159/pkg/controller/cronjob/cronjob_controllerv2.go#L498-L505) you can safely log that we fallback to kube-controller-manager's timezone and do nothing, but since this method invocations is further down the chain after that failures, it should not reach it.

**Review by soltysh — CHANGES_REQUESTED** [review]
/lgtm cancel

File: pkg/controller/cronjob/utils.go
Comment (liggitt) [inline pkg/controller/cronjob/utils.go:232]: I don't think you need to do any of the timezone stuff here, isn't the scheduledTime returned from the cron scheduler already in the configured timezone, either the explicit one or the controller-manager's local time?

File: staging/src/k8s.io/api/batch/v1/types.go
Comment (liggitt) [inline staging/src/k8s.io/api/batch/v1/types.go:31]: this description doesn't seem accurate... it's not recording the creation timestamp, it's recording the scheduled timestamp

File: pkg/controller/cronjob/utils.go
Comment (helayoty) [reply pkg/controller/cronjob/utils.go]: The Timezone section has been removed and also fixed the annotation issue.

File: staging/src/k8s.io/api/batch/v1/types.go
Comment (helayoty) [reply staging/src/k8s.io/api/batch/v1/types.go]: Addressed.

File: pkg/features/kube_features.go
Comment (helayoty) [reply pkg/features/kube_features.go:744]: Addressed.

File: staging/src/k8s.io/api/batch/v1/types.go
Comment (helayoty) [reply staging/src/k8s.io/api/batch/v1/types.go]: Addressed.

File: pkg/features/kube_features.go
Comment (liggitt) [inline pkg/features/kube_features.go:1046]: alphabetize here as well

File: staging/src/k8s.io/api/batch/v1/types.go
Comment (liggitt) [inline staging/src/k8s.io/api/batch/v1/types.go:33]: this looks good.. open a PR to the dev-1.28 branch of the website to add this doc to https://kubernetes.io/docs/reference/labels-annotations-taints/ as well

**Review by liggitt — COMMENTED** [review]
one nit on the feature gate order, then lgtm

**Comment by liggitt** [comment]
/approve
for API constant addition

/hold
for feature gate reorder and squash

@soltysh can lgtm and unhold once that is done

File: staging/src/k8s.io/api/batch/v1/types.go
Comment (helayoty) [reply staging/src/k8s.io/api/batch/v1/types.go:33]: PR is ready already. https://github.com/kubernetes/website/pull/41857

File: pkg/features/kube_features.go
Comment (helayoty) [reply pkg/features/kube_features.go]: Addressed.

File: pkg/controller/cronjob/utils.go
Comment (soltysh) [inline pkg/controller/cronjob/utils.go:219]: This is still wrong, it should be:
```golang
annotations[batchv1.CronJobScheduledTimestampAnnotation] = scheduledTime.Format(time.RFC3339).In(timezone) 
```
this is because we want to present that annotation in the time that is local to the user, and the information in schedule is transformed to current time zone.

**Review by soltysh — COMMENTED** [review]
/lgtm cancel

File: pkg/controller/cronjob/utils.go
Comment (soltysh) [reply pkg/controller/cronjob/utils.go:219]: It's hard to quickly explain, but when we invoke `cron.ParseStandard` we pass `TZ=<time-zone` which triggers the library to take into account the timezone when calculating the time, but the date returned from invoking `Next` on that schedule will give you time relative to your current time zone. In our case that is the time zone local to the kube-controller-manager. 

It's best to try it locally with something like this:
```golang
sched, _ := cron.ParseStandard("TZ=America/New_York 0 */4 * * *")
next := sched.Next(time.Now())
loc, _ := time.LoadLocation("America/New_York")
fmt.Println("next in TZ", next.In(loc))
fmt.Println("next in local", next)
```
make sure to pick the time zone that is different than your local, the schedule is set to execute every 4h at 0 minutes. 
In my case the above will print (it's currently 21:04pm CEST, which is 15:05 EST):
```
next in TZ 2023-07-06 16:00:00 -0400 EDT
next in local 2023-07-06 22:00:00 +0200 CEST
```
for the annotation we want to present the user with the time printed in TZ.

Comment (helayoty) [reply pkg/controller/cronjob/utils.go]: Addressed.

**Review by soltysh — APPROVED** [review]
/lgtm

File: pkg/controller/cronjob/utils.go
Comment (liggitt) [reply pkg/controller/cronjob/utils.go]: > the date returned from invoking Next on that schedule will give you time relative to your current time zone

ah, I see, it uses the location of the time passed in to Next()

**Pull request merged**
